### PR TITLE
Add support for depending on shared libraries on linux.

### PR DESCRIPTION
**Pull request by mfarrugi**
This change allows `rust_library` and `rust_binary` to depend on native shared libraries.
This requires 2 additional steps, collecting shared libraries from cc_library and rust_library, and setting the rpath in executables.

I added an example of this, and tested more extensively on another project.

I am not sure how much work is left to support mac+windows or rust dylibs.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by mfarrugi** [comment]
@damienmg can someone take a look at this? I'm not that confident this is how it should be done.

**Comment by damienmg** [comment]
+David and Marcel

**Comment by mfarrugi** [comment]
@davidzchen? 

Would love any expert bazel eyes on this.

File: rust/rust.bzl [hunk 1]
```diff
@@ -49,11 +49,12 @@ rust_repositories()
   [Cargo](https://crates.io/).
 """
 
-load(":toolchain.bzl", "build_rustc_command", "build_rustdoc_command", "build_rustdoc_test_command")
+load(":toolchain.bzl", "build_rustc_command", "build_rustdoc_command", "build_rustdoc_test_command", "relative_path")
```
Comment (davidzchen) [inline rust/rust.bzl]: Keep line length to 80 chars.

File: rust/toolchain.bzl [hunk 2]
```diff
@@ -101,6 +106,19 @@ def build_rustdoc_test_command(ctx, toolchain, depinfo, lib_rs):
       depinfo.search_flags +
       depinfo.link_flags)
 
+# @TODO(review) How should this be done? cc_* logic seems much more involved.
```
Comment (davidzchen) [inline rust/toolchain.bzl]: @dslomov Can you help with this portion of the review? Thanks.

File: rust/toolchain.bzl [hunk 3]
```diff
@@ -133,6 +151,39 @@ def _get_files(input):
...
+  Args:
+    path_parts: A list containing parts of a path.
+
+  Returns:
+    Returns a list containing the path parts with all "." elements removed.
+  """
+  path_parts = path.split("/")
+  return [part for part in path_parts if part != "."]
+
+def relative_path(src_path, dest_path):
```
Comment (davidzchen) [inline rust/toolchain.bzl]: If this function is internal to this bzl file, prefix with `_`

Comment (mfarrugi) [reply rust/toolchain.bzl]: It's also used in rust/rust.bzl

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: rust/toolchain.bzl (hunk 3 above)
Comment (davidzchen) [reply rust/toolchain.bzl]: Ah you're right.

File: rust/toolchain.bzl (hunk 2 above)
Comment (davidzchen) [reply rust/toolchain.bzl]: @mhlopko Can you help take a look at this part of the review?

Comment (mfarrugi) [reply rust/toolchain.bzl]: In particular, I'm concerned about handling the placement of .so's emitted by setting crate_type=dylib (which is supported by another PR). 

For depending on cc_* rule dylibs I don't think there is any concern, but the cc_* logic is suspiciously complex.

**Comment by acmcarther** [comment]
Added @dslomov and @mhlopko to get eyes on the toolchain.bzl changes

**Comment by mfarrugi** [comment]
How do I get CI to run on this? Maybe would have caught https://github.com/acmcarther/cargo-raze/issues/31

**Comment by acmcarther** [comment]
ok to test

File: rust/toolchain.bzl [hunk 4]
```diff
@@ -23,11 +23,15 @@ def build_rustc_command(ctx, toolchain, crate_name, crate_type, src, output_dir,
...
 
+  rpaths = _compute_rpaths(ctx.bin_dir, output_dir, depinfo)
+
   # Construct features flags
   features_flags = _get_features_flags(ctx.attr.crate_features)
 
   return " ".join(
       ["set -e;"] +
+      # If TMPDIR is set but not created, rustc will die.
+      ["if [[ -v TMPDIR ]]; then mkdir -p $TMPDIR; fi;"] +
```
Comment (acmcarther) [inline rust/toolchain.bzl]: "-v" is apparently only available since bash 4.2, and "[[ ]]" is apparently not portable.

Try "if [ ! -z "${TMPDIR+x}" ];" ?

**Comment by mfarrugi** [comment]
CI requires this to work for osx (but not windows?), which requires a very different approach to setting rpath. I would appreciate some bazel-er input before reinventing how cpp linking is done in skylark.

cc https://github.com/acmcarther/cargo-raze/issues/27 I stand surprised :)

**Comment by mfarrugi** [comment]
Going to need to skip the osx tests for this, not sure how to do that yet.

**Comment by acmcarther** [comment]
I've been building from under a rock apparently. This change is what makes depending on system libraries work! I've been silently using this in my side project, and when I went to use bazelbuild/rules_rust:master, everything broke e.g:
```
error: could not find native static library `clang`, perhaps an -L flag is missing?
```

mfarrugi@: can you remind me what the next steps are on this PR? From what I can tell:

1) Fix up merge conflicts
2) Another pass of code review
3) Ideally, a solution that makes builds work on Windows

I don't have the best background for (2) and (3), but I can read up enough for a passable code review and maybe help come up with something functional enough for now.

**Comment by mfarrugi** [comment]
The implementation on windows and osx are both completely different, which I would also have to read up on in order to do. I'm wary of re-implementing this logic when there aren't (many?) people asking for it and more generic dylib linking is ostensibly in the pipeline as part of the cc skylark project.

**Comment by bobsomers** [comment]
This is a good patch, and I personally don't care about proper MacOS or Windows support for this, proper linking with shared libraries on Linux is good enough! 🎉

**Comment by mfarrugi** [comment]
@acmcarther is linux support enough? If so, this has been merged and is ready for review.

**Comment by acmcarther** [comment]
I'm agreeable to linux-only support for this, but it ought to fail cleanly when attempting to use the feature outside of a linux context until we can get it working in those other contexts. That needs some thought, but a hack based on information available in the toolchain, and calling `fail(...)` as is done in other circumstances is probably sufficient.

You'll want to sync to master here as well.

Given that this is a load bearing PR (my repos, for one depend on it), lets get this reviewed and submitted even if others can't come by. I'll go for another real thorough pass in the next 12ish hours and then we should be good. Thanks for this important change, and your patience on the slow review process.

File: rust/rust.bzl [hunk 5]
```diff
@@ -63,38 +62,10 @@ HTML_MD_FILETYPE = FileType([
...
-  relative_path = ""
-  for i in range(n, len(src_parts)):
-    relative_path += "../"
-  relative_path += "/".join(dest_parts[n:])
-
-  return relative_path
+def _get_lib_name(lib):
+  """ Returns the name of a library artifact, eg. libabc.a -> abc"""
+  libname, ext = lib.basename.split(".", 2)
+  return libname[3:]
```
Comment (acmcarther) [inline rust/rust.bzl]: Is it worth checking to make sure that `lib.startswith("lib")`?

File: rust/toolchain.bzl (hunk 2 above)
Comment (acmcarther) [reply rust/toolchain.bzl]: Did your concerns about this get resolved? Either way, we probably don't want to submit this with this TODO and below comment intact.

**Comment by acmcarther** [comment]
I'm mostly caught up at this point on the way this works. The dep tracking strikes me as sloppy (in particular, the way we're tracking [libs, transitive_libs, transitive_dylibs, and transitive_staticlibs]), but I suspect that might just be my unfamiliarity with the problem space.

I've got a couple of comments mostly around error handling and being a little more explicit about assumptions, and I'm curious what the resolution to the `cc_*` confusion was,  but overall it looks good.

**Comment by bobsomers** [comment]
> The dep tracking strikes me as sloppy (in particular, the way we're tracking [libs, transitive_libs, transitive_dylibs, and transitive_staticlibs]), but I suspect that might just be my unfamiliarity with the problem space.

It is kinda sloppy, but a proper fix for #78 will likely end up rewriting a lot of that to clean it up (at least from some preliminary hacking I have in a local branch).

IMHO it probably makes sense to keep it this way for now, since it's a logical extension of how it already works, and then clean it all up in one go when a fix for #78 is tackled.

(Sorry for the drive-by review comments, when I get some free time I'm planning to take a whack at solving #78 and it conflicts enough with this that one of them needs to go first.)

File: rust/toolchain.bzl [hunk 6]
```diff
@@ -110,6 +115,27 @@ def build_rustdoc_test_command(ctx, toolchain, depinfo, lib_rs):
...
+# 2. cc sets both RUNPATH and RPATH in certain circumstances.
+def _compute_rpaths(toolchain, bin_dir, output_dir, depinfo):
+  """
+  Determine the artifact's rpath relative to the bazel root.
+  """
+  if not depinfo.transitive_dylibs:
+    return []
+
+  if toolchain.os != 'linux':
+    print("Runtime linking is not supported on {}, but found {}".format(
```
[the remark this answers is not in the document]
Comment (mfarrugi) [reply rust/toolchain.bzl]: This builds correctly, but run will fall. My thinking was that this technically works, and is easier to iterate on for users and contributors.

File: rust/toolchain.bzl [hunk 7]
```diff
@@ -142,6 +163,39 @@ def _get_files(input):
...
+  Args:
+    path_parts: A list containing parts of a path.
+
+  Returns:
+    Returns a list containing the path parts with all "." elements removed.
+  """
+  path_parts = path.split("/")
+  return [part for part in path_parts if part != "."]
+
+def relative_path(src_path, dest_path):
```
[the remark this answers is not in the document]
Comment (mfarrugi) [reply rust/toolchain.bzl]: I think there are enough of these utility methods to pull out a utils.bzl, should I hold off on that?

File: rust/toolchain.bzl (hunk 6 above)
[the remark this answers is not in the document]
Comment (mfarrugi) [reply rust/toolchain.bzl]: CI says I'm wrong, so I'll fail here and ignore it in ci builds. (https://source.cloud.google.com/results/invocations/46ee1fc3-222d-491f-bbb7-24d208119f66/log)

File: rust/toolchain.bzl (hunk 2 above)
Comment (mfarrugi) [reply rust/toolchain.bzl]: No, but I haven't had issues and neither has anyone else, so that's good enough for now.

**Comment by mfarrugi** [comment]
@acmcarther this is up to date again

**Comment by acmcarther** [comment]
```
rule @examples//hello_lib:hello_lib doesn't provide attribute transitive_libs. Its list of attributes is: ["actions", "crate_root", "crate_type", "data_runfiles", "default_runfiles", "files", "files_to_run", "label", "output_group", "output_groups", "rust_deps", "rust_lib", "rust_srcs", "transitive_dylibs", "transitive_rlibs", "transitive_staticlibs"]
```

**Comment by mfarrugi** [comment]
@acmcarther sorry about that, was running against `@examples` and not `...`.

Can I get added to the auto-ci whitelist to streamline changes a little bit more?

**Comment by mfarrugi** [comment]
@acmcarther do we want anything else here? If so, lmk and we can let #84 go ahead next while I work that out (cc @bobsomers ).

**Comment by acmcarther** [comment]
Regd https://github.com/bazelbuild/rules_rust/pull/61#issuecomment-382410499

I don't actually know how to add someone to the whitelist and https://github.com/bazelbuild/continuous-integration/blob/master/buildkite/README.md#pull-requests wasn't immediately enlightening.

@buchgr: What's the org-wide attitude toward adding people to the CI whitelist? Is there a suggested approach to doing that, or perhaps is "ci whitelist" considered roughly equivalent to maintainer status anyway?

---

Regd https://github.com/bazelbuild/rules_rust/pull/61#issuecomment-390834671

This looks good to me, but there is a merge conflict. Can you fix that and tag me again and I'll ship.

**Comment by buchgr** [comment]
@acmcarther you can add someone to the whitelist by adding him as a collaborator of rules_rust on Github.

**Comment by buchgr** [comment]
I have added @mfarrugi as a collaborator (Settings -> Teams & Collaborators). Need to accept the invite and then CI should just work.

**Comment by mfarrugi** [comment]
@acmcarther merged, not sure what exactly bot is unhappy with though.

**Comment by acmcarther** [comment]
https://github.com/bazelbuild/rules_rust/pull/61#issuecomment-391446489
Merge included 634a6f3d7e35c98fe8452d7cddb85ca1f0662d5d which is already merged to master. CLAs all signed.
@mfarrugi: I think the workflow expects you to rebase, not merge master, so that the upstream commits aren't directly included here. No problem though.

---

@buchgr: Thank you for the quick attention! I'll remember the collaborator process and use it in the future.

**Pull request merged**